### PR TITLE
GRPC Update Aggregates [CHA-3925]

### PR DESCRIPTION
**Pull request by rooftoofwoof**

### CODE REVIEW

File: test_fixtures.go [hunk 1]
```diff
@@ -49,7 +49,36 @@ type dclassWithOverrides struct {
 	CamelName *string `dataclass_field:"true" name:"camelName"`
 }
 
-type allTypes struct {
+
+
+type allTypesInput struct {
```
Comment (rooftoofwoof) [inline test_fixtures.go:54]: remove

File: errors.go [hunk 2]
```diff
@@ -67,3 +68,16 @@ func deferFunctionWithError(function func() error, originalError error) error {
...
+	return &ErrorResponse{ServerErrors: errs}
+}
+
+func newClientError(err error) *ErrorResponse {
+	return &ErrorResponse{ClientError: &ClientError{Message: err.Error()}}
+}
+
+func clientWrap(err error, message string) *ErrorResponse {
+	wrapped := errors.Wrap(err, message)
+	return newClientError(wrapped)
```
Comment (rooftoofwoof) [inline errors.go:82]: inline

File: params_upload.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			}
+			fqn = feature.Fqn
+		}
+		res[fqn] = v
+
+		currLength := -1
+		if reflect.TypeOf(v).Kind() == reflect.Slice || reflect.TypeOf(v).Kind() == reflect.Array {
+			currLength = reflect.ValueOf(v).Len()
+		} else {
+			return nil, &ErrorResponse{
```
Comment (rooftoofwoof) [inline params_upload.go:31]: return plain errors

File: client_grpc.go [hunk 4]
```diff
@@ -452,7 +446,36 @@ func (c *clientGrpc) OnlineQueryBulk(args OnlineQueryParamsComplete) (OnlineQuer
...
+	if err != nil {
+		return UploadFeaturesResult{}, wrapClientError(err, "error serializing inputs as feather")
+	}
+
+	req := connect.NewRequest(&commonv1.UploadFeaturesBulkRequest{
+		InputsFeather: inputsFeather,
+		BodyType:      commonv1.FeatherBodyType_FEATHER_BODY_TYPE_TABLE,
+	})
+
+	res, err := c.queryClient.UploadFeaturesBulk(context.Background(), req)
```
Comment (ravirahman) [inline client_grpc.go:463]: is this the right context to use here? Should the caller be able to pass one in? Not sure what is the common thing to do in golang

File: internal/feather.go [hunk 5]
```diff
@@ -176,31 +178,36 @@ func setBuilderValues(builder array.Builder, slice reflect.Value, valid []bool)
...
+				return errors.Errorf(
+					"expected slice of slices, instead found slice of '%s'",
+					innerSliceType.Kind(),
+				)
+			}
+			flatSlice := reflect.MakeSlice(innerSliceType, 0, slice.Len())
+			highwaterMark := 0
+			offsets = append(offsets, 0)
+			for i := 0; i < slice.Len(); i++ {
+				innerSlice := slice.Index(i)
```
Comment (ravirahman) [inline internal/feather.go:197]: Does this handle nulls correctly?

File: client_grpc.go (hunk 4 above)
Comment (rooftoofwoof) [reply client_grpc.go:463]: good question - idk either let me go look around

File: internal/feather.go [hunk 6]
```diff
@@ -176,31 +178,36 @@ func setBuilderValues(builder array.Builder, slice reflect.Value, valid []bool)
...
+				flatSlice = reflect.AppendSlice(flatSlice, innerSlice)
+			}
+			if err := setBuilderValues(
+				builder.(*array.LargeListBuilder).ValueBuilder(),
+				flatSlice,
+				allValid(flatSlice.Len()),
+			); err != nil {
+				return errors.Wrap(err, "failed to set values for slice of slices")
+			}
+			builder.(*array.LargeListBuilder).AppendValues(offsets, valid)
```
Comment (ravirahman) [inline internal/feather.go:209]: justa nna confirm that we control that this is a large list builder and not say a list builder

File: internal/feather.go [hunk 7]
```diff
@@ -315,7 +323,56 @@ func ColumnMapToRecord(inputs map[string]any) (arrow.Record, error) {
...
+	var tsField arrow.Field
+	for idx, field := range res.Schema().Fields() {
+		if field.Name == "__ts__" {
+			tsIndex = idx
+			tsField = field
+			break
+		}
+	}
+	if tsIndex != -1 {
+		defer res.Release()
```
Comment (ravirahman) [inline internal/feather.go:342]: should we always defer this release -- i.e. move to line 328?

File: internal/feather.go [hunk 8]
```diff
@@ -315,7 +323,56 @@ func ColumnMapToRecord(inputs map[string]any) (arrow.Record, error) {
...
+		defer tsBuilder.Release()
+		tsValues := make([]arrow.Timestamp, res.NumRows())
+		validityMask := make([]bool, res.NumRows())
+		for i := 0; i < int(res.NumRows()); i++ {
+			validityMask[i] = res.Column(tsIndex).IsValid(i)
+			tsCol, ok := res.Column(tsIndex).(*array.Timestamp)
+			if !ok {
+				return nil, fmt.Errorf("failed to cast '__ts__' column to Timestamp Array")
+			}
+			val := tsCol.Value(i)
```
Comment (ravirahman) [inline internal/feather.go:353]: is it valid to get the value if isValid() return false? Or is it expected that all invalid values are zero-constructed?

**Review by ravirahman — APPROVED** [review]
Not too famaliar but overall lgtm; see comments mainly re null handling

File: internal/feather.go (hunk 7 above)
Comment (rooftoofwoof) [reply internal/feather.go:342]: My intent here is to defer Release of the old record with uncast the column, and return a non-defer-released new record with the cast column

File: internal/feather.go (hunk 8 above)
Comment (rooftoofwoof) [reply internal/feather.go:353]: Yes - unless the method returns an error, it shouldn't panic/go into an error state in any case

Comment (rooftoofwoof) [reply internal/feather.go:353]: but good caution

File: client_grpc.go (hunk 4 above)
Comment (rooftoofwoof) [reply client_grpc.go:463]: addressed

File: internal/feather.go (hunk 6 above)
Comment (rooftoofwoof) [reply internal/feather.go:209]: yes, thank you!

File: client.go [hunk 9]
```diff
@@ -108,6 +108,31 @@ type Client interface {
 	// [chalk codegen]: https://docs.chalk.ai/cli#codegen
 	UploadFeatures(args UploadFeaturesParams) (UploadFeaturesResult, error)
 
+	// UpdateAggregates synchronously persists feature values that backs windowed aggregations.
```
Comment (rooftoofwoof) [inline client.go:111]: "back", no "s"

File: internal/feather.go [hunk 10]
```diff
@@ -315,7 +323,56 @@ func ColumnMapToRecord(inputs map[string]any) (arrow.Record, error) {
...
+	tsIndex := -1
+	var tsField arrow.Field
+	for idx, field := range res.Schema().Fields() {
+		if field.Name == "__ts__" {
+			tsIndex = idx
+			tsField = field
+			break
+		}
+	}
+	if tsIndex != -1 {
```
Comment (rooftoofwoof) [inline internal/feather.go:341]: this is possibly made redundant with Raphael's change in the backend to cast colujmns accordingly

Comment (rooftoofwoof) [reply internal/feather.go]: removed

File: client.go (hunk 9 above)
Comment (rooftoofwoof) [reply client.go]: fixed

**Pull request merged**